diff --git a/src/libxl/libxl_conf.c b/src/libxl/libxl_conf.c
index 00748e21e8..81665dd7b1 100644
--- a/src/libxl/libxl_conf.c
+++ b/src/libxl/libxl_conf.c
@@ -931,6 +931,7 @@ libxlMakeNetworkDiskSrcStr(virStorageSourcePtr src,
 
     switch ((virStorageNetProtocol) src->protocol) {
     case VIR_STORAGE_NET_PROTOCOL_NBD:
+    case VIR_STORAGE_NET_PROTOCOL_CBD:
     case VIR_STORAGE_NET_PROTOCOL_HTTP:
     case VIR_STORAGE_NET_PROTOCOL_HTTPS:
     case VIR_STORAGE_NET_PROTOCOL_FTP:
diff --git a/src/libxl/xen_xl.c b/src/libxl/xen_xl.c
index ed0ce77db4..aea2eee621 100644
--- a/src/libxl/xen_xl.c
+++ b/src/libxl/xen_xl.c
@@ -1590,6 +1590,7 @@ xenFormatXLDiskSrcNet(virStorageSourcePtr src)
 
     switch ((virStorageNetProtocol) src->protocol) {
     case VIR_STORAGE_NET_PROTOCOL_NBD:
+    case VIR_STORAGE_NET_PROTOCOL_CBD:
     case VIR_STORAGE_NET_PROTOCOL_HTTP:
     case VIR_STORAGE_NET_PROTOCOL_HTTPS:
     case VIR_STORAGE_NET_PROTOCOL_FTP:
diff --git a/src/qemu/qemu_block.c b/src/qemu/qemu_block.c
index 4640e339c0..0100f3c8ee 100644
--- a/src/qemu/qemu_block.c
+++ b/src/qemu/qemu_block.c
@@ -881,6 +881,18 @@ qemuBlockStorageSourceGetRBDProps(virStorageSourcePtr src,
     return ret;
 }
 
+static virJSONValuePtr
+qemuBlockStorageSourceGetCBDProps(virStorageSourcePtr src)
+{
+    virJSONValuePtr ret = NULL;
+    g_autofree char *str = NULL;
+
+    str = g_strdup_printf("cbd:%s", src->path);
+    ignore_value(virJSONValueObjectCreate(&ret,
+                                          "s:filename", str,
+                                          NULL) < 0);
+    return ret;
+}
 
 static virJSONValuePtr
 qemuBlockStorageSourceGetSheepdogProps(virStorageSourcePtr src)
@@ -1168,6 +1180,12 @@ qemuBlockStorageSourceGetBackendProps(virStorageSourcePtr src,
                 return NULL;
             break;
 
+        case VIR_STORAGE_NET_PROTOCOL_CBD:
+            driver = "cbd";
+            if (!(fileprops = qemuBlockStorageSourceGetCBDProps(src)))
+                return NULL;
+            break;
+
         case VIR_STORAGE_NET_PROTOCOL_SHEEPDOG:
             driver = "sheepdog";
             if (!(fileprops = qemuBlockStorageSourceGetSheepdogProps(src)))
@@ -2110,6 +2128,7 @@ qemuBlockGetBackingStoreString(virStorageSourcePtr src,
 
             case VIR_STORAGE_NET_PROTOCOL_SHEEPDOG:
             case VIR_STORAGE_NET_PROTOCOL_RBD:
+            case VIR_STORAGE_NET_PROTOCOL_CBD:
             case VIR_STORAGE_NET_PROTOCOL_VXHS:
             case VIR_STORAGE_NET_PROTOCOL_SSH:
             case VIR_STORAGE_NET_PROTOCOL_LAST:
@@ -2508,6 +2527,7 @@ qemuBlockStorageSourceCreateGetStorageProps(virStorageSourcePtr src,
         case VIR_STORAGE_NET_PROTOCOL_FTPS:
         case VIR_STORAGE_NET_PROTOCOL_TFTP:
         case VIR_STORAGE_NET_PROTOCOL_NONE:
+        case VIR_STORAGE_NET_PROTOCOL_CBD:
         case VIR_STORAGE_NET_PROTOCOL_LAST:
             return 0;
         }
diff --git a/src/qemu/qemu_command.c b/src/qemu/qemu_command.c
index 479bcc0b0c..30eb879d3e 100644
--- a/src/qemu/qemu_command.c
+++ b/src/qemu/qemu_command.c
@@ -1018,6 +1018,12 @@ qemuBuildNetworkDriveStr(virStorageSourcePtr src,
                            _("VxHS protocol does not support URI syntax"));
             return NULL;
 
+        case VIR_STORAGE_NET_PROTOCOL_CBD:
+            virBufferStrcat(&buf, "cbd:", src->path, NULL);
+
+            ret = virBufferContentAndReset(&buf);
+            break;
+
         case VIR_STORAGE_NET_PROTOCOL_SSH:
             virReportError(VIR_ERR_INTERNAL_ERROR, "%s",
                            _("'ssh' protocol is not yet supported"));
diff --git a/src/qemu/qemu_domain.c b/src/qemu/qemu_domain.c
index 663c0af867..74ec2c4703 100644
--- a/src/qemu/qemu_domain.c
+++ b/src/qemu/qemu_domain.c
@@ -9580,6 +9580,7 @@ qemuDomainPrepareStorageSourceTLS(virStorageSourcePtr src,
         break;
 
     case VIR_STORAGE_NET_PROTOCOL_RBD:
+    case VIR_STORAGE_NET_PROTOCOL_CBD:
     case VIR_STORAGE_NET_PROTOCOL_SHEEPDOG:
     case VIR_STORAGE_NET_PROTOCOL_GLUSTER:
     case VIR_STORAGE_NET_PROTOCOL_ISCSI:
diff --git a/src/qemu/qemu_snapshot.c b/src/qemu/qemu_snapshot.c
index 15494c3415..62992035e8 100644
--- a/src/qemu/qemu_snapshot.c
+++ b/src/qemu/qemu_snapshot.c
@@ -403,6 +403,7 @@ qemuSnapshotPrepareDiskExternalInactive(virDomainSnapshotDiskDefPtr snapdisk,
         case VIR_STORAGE_NET_PROTOCOL_NONE:
         case VIR_STORAGE_NET_PROTOCOL_NBD:
         case VIR_STORAGE_NET_PROTOCOL_RBD:
+        case VIR_STORAGE_NET_PROTOCOL_CBD:
         case VIR_STORAGE_NET_PROTOCOL_SHEEPDOG:
         case VIR_STORAGE_NET_PROTOCOL_GLUSTER:
         case VIR_STORAGE_NET_PROTOCOL_ISCSI:
@@ -492,6 +493,7 @@ qemuSnapshotPrepareDiskExternalActive(virDomainObjPtr vm,
         case VIR_STORAGE_NET_PROTOCOL_NONE:
         case VIR_STORAGE_NET_PROTOCOL_NBD:
         case VIR_STORAGE_NET_PROTOCOL_RBD:
+        case VIR_STORAGE_NET_PROTOCOL_CBD:
         case VIR_STORAGE_NET_PROTOCOL_SHEEPDOG:
         case VIR_STORAGE_NET_PROTOCOL_ISCSI:
         case VIR_STORAGE_NET_PROTOCOL_HTTP:
@@ -621,6 +623,7 @@ qemuSnapshotPrepareDiskInternal(virDomainDiskDefPtr disk,
         case VIR_STORAGE_NET_PROTOCOL_NONE:
         case VIR_STORAGE_NET_PROTOCOL_NBD:
         case VIR_STORAGE_NET_PROTOCOL_RBD:
+        case VIR_STORAGE_NET_PROTOCOL_CBD:
         case VIR_STORAGE_NET_PROTOCOL_SHEEPDOG:
         case VIR_STORAGE_NET_PROTOCOL_GLUSTER:
         case VIR_STORAGE_NET_PROTOCOL_ISCSI:
diff --git a/src/util/virstoragefile.c b/src/util/virstoragefile.c
index 35e6dcf5de..660866adc8 100644
--- a/src/util/virstoragefile.c
+++ b/src/util/virstoragefile.c
@@ -78,6 +78,7 @@ VIR_ENUM_IMPL(virStorageFileFeature,
 VIR_ENUM_IMPL(virStorageNetProtocol,
               VIR_STORAGE_NET_PROTOCOL_LAST,
               "none",
+              "cbd",
               "nbd",
               "rbd",
               "sheepdog",
@@ -3151,6 +3152,7 @@ virStorageSourceParseBackingColon(virStorageSourcePtr src,
     case VIR_STORAGE_NET_PROTOCOL_ISCSI:
     case VIR_STORAGE_NET_PROTOCOL_GLUSTER:
     case VIR_STORAGE_NET_PROTOCOL_SSH:
+    case VIR_STORAGE_NET_PROTOCOL_CBD:
     case VIR_STORAGE_NET_PROTOCOL_VXHS:
         virReportError(VIR_ERR_INTERNAL_ERROR,
                        _("malformed backing store path for protocol %s"),
@@ -3723,6 +3725,15 @@ virStorageSourceParseBackingJSONRBD(virStorageSourcePtr src,
     return 0;
 }
 
+static int
+virStorageSourceParseBackingJSONCBD(virStorageSourcePtr src G_GNUC_UNUSED,
+                                    virJSONValuePtr json G_GNUC_UNUSED,
+                                    const char *jsonstr G_GNUC_UNUSED,
+                                    int opaque G_GNUC_UNUSED)
+{
+    return -1;//not support
+}
+
 static int
 virStorageSourceParseBackingJSONRaw(virStorageSourcePtr src,
                                     virJSONValuePtr json,
@@ -3853,6 +3864,7 @@ static const struct virStorageSourceJSONDriverParser jsonParsers[] = {
     {"sheepdog", false, virStorageSourceParseBackingJSONSheepdog, 0},
     {"ssh", false, virStorageSourceParseBackingJSONSSH, 0},
     {"rbd", false, virStorageSourceParseBackingJSONRBD, 0},
+    {"cbd", false, virStorageSourceParseBackingJSONCBD, 0},
     {"raw", true, virStorageSourceParseBackingJSONRaw, 0},
     {"vxhs", false, virStorageSourceParseBackingJSONVxHS, 0},
     {"nvme", false, virStorageSourceParseBackingJSONNVMe, 0},
@@ -4629,6 +4641,7 @@ virStorageSourceNetworkDefaultPort(virStorageNetProtocol protocol)
 
         case VIR_STORAGE_NET_PROTOCOL_LAST:
         case VIR_STORAGE_NET_PROTOCOL_NONE:
+        case VIR_STORAGE_NET_PROTOCOL_CBD:
             return 0;
     }
 
diff --git a/src/util/virstoragefile.h b/src/util/virstoragefile.h
index 87763cf389..91beb2b698 100644
--- a/src/util/virstoragefile.h
+++ b/src/util/virstoragefile.h
@@ -123,6 +123,7 @@ struct _virStorageTimestamps {
 /* Information related to network storage */
 typedef enum {
     VIR_STORAGE_NET_PROTOCOL_NONE,
+    VIR_STORAGE_NET_PROTOCOL_CBD,
     VIR_STORAGE_NET_PROTOCOL_NBD,
     VIR_STORAGE_NET_PROTOCOL_RBD,
     VIR_STORAGE_NET_PROTOCOL_SHEEPDOG,
